src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -201,28 +202,45 @@ public void phase1(final SnapshotIndexCommit snapshot, final Translog.View trans
             // Generate a "diff" of all the identical, different, and missing
             // segment files on the target node, using the existing files on
             // the source node
-            final Store.RecoveryDiff diff = recoverySourceMetadata.recoveryDiff(new Store.MetadataSnapshot(request.existingFiles()));
-            for (StoreFileMetaData md : diff.identical) {
-                response.phase1ExistingFileNames.add(md.name());
-                response.phase1ExistingFileSizes.add(md.length());
-                existingTotalSize += md.length();
-                if (logger.isTraceEnabled()) {
-                    logger.trace("[{}][{}] recovery [phase1] to {}: not recovering [{}], exists in local store and has checksum [{}], size [{}]",
-                            indexName, shardId, request.targetNode(), md.name(), md.checksum(), md.length());
+            String recoverySourceSyncId = recoverySourceMetadata.getSyncId();
+            String recoveryTargetSyncId = request.metadataSnapshot().getSyncId();
+            final boolean recoverWithSyncId = recoverySourceSyncId != null &&
+                    recoverySourceSyncId.equals(recoveryTargetSyncId);
+            if (recoverWithSyncId) {
+                for (StoreFileMetaData md : request.metadataSnapshot()) {
+                    response.phase1ExistingFileNames.add(md.name());
```

Comment:
@brwe I looked into this again and we discussed this briefly before. I think we should add this

```Java
  /**
     * Returns the number of document in the index referenced by this {@link SegmentInfos}
     */
    public static int getNumDocs(SegmentInfos info) {
        int numDocs = 0;
        for (SegmentCommitInfo si : info) {
            numDocs +=si.info.maxDoc() - si.getDelCount();
        }
        return numDocs;
    }
```

to `Lucene.java` as a util and add the value to `Store.MetadataSnapshot` and transfer it with our request before we do the phase 1 comparison. That way we can simply compare the number and fail everything if they don't match?